language/tools/move-package/src/resolution/resolution_graph.rs

```diff
@@ -0,0 +1,617 @@
+                bail!(
+                    "Duplicate renaming of named address '{}' found in dependency '{}'",
+                    rename_to,
+                    dep_name
+                );
+            }
+        }
+        Ok(())
+    }
+
+    // The resolution table contains the transitive closure of addresses that are known in that
+    // package. Extends the package's resolution table and checks for duplicate renamings that
+    // conflict during this process.
+    fn extend_resolution_table(
+        resolution_table: &mut ResolvingTable,
+        dep_name: &PackageName,
+        dep_resolution_table: ResolvingTable,
+        dep_renaming: Renaming,
+    ) -> Result<()> {
+        let renames = dep_renaming
+            .into_iter()
+            .map(|(rename_to, (_, rename_from))| (rename_from, rename_to))
+            .collect::<BTreeMap<_, _>>();
+
+        for (addr_name, addr_value) in dep_resolution_table.into_iter() {
+            let addr_name = if let Some(rename_into) = renames.get(&addr_name) {
+                rename_into.clone()
+            } else {
+                addr_name
+            };
+            if let Some(other) = resolution_table.insert(addr_name.clone(), addr_value.clone()) {
+                // They need to be the same refcell so resolve to the same location if there are any
+                // possible reassignments
+                if other.value != addr_value.value {
+                    bail!(
+                        "Named address '{}' in dependency '{}' is already set to '{}' but was then reassigned to '{}'",
+                        &addr_name,
+                        dep_name,
+                        match other.value.take() {
+                            None => "unassigned".to_string(),
+                            Some(addr) => format!("0x{}", addr.short_str_lossless()),
+                        },
+                        match addr_value.value.take() {
+                            None => "unassigned".to_string(),
+                            Some(addr) => format!("0x{}", addr.short_str_lossless()),
+                        }
+                    );
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl ResolvingNamedAddress {
+    pub fn new(address_opt: Option<AccountAddress>) -> Self {
+        Self {
+            value: Rc::new(RefCell::new(address_opt)),
+        }
+    }
+
+    pub fn unify(&self, address_opt: Option<AccountAddress>) -> Result<()> {
+        match address_opt {
+            None => Ok(()),
+            Some(addr_val) => match self.value.take() {
+                None => {
+                    self.value.replace(Some(addr_val));
+                    Ok(())
+                }
+                Some(current_value) => {
+                    if current_value != addr_val {
+                        bail!("Attempted to assign a different value '0x{}' to an a already-assigned named address '0x{}'",
```

Comment:
The main issue is the setting of the value in the `None` case since if we borrow we can't perform a `replace` on the value. I've updated it so that we don't need to do a replace though (and we do a mutable borrow). So let me know if you like the upcoming change better.